### PR TITLE
Boilerplate refactor

### PR DESCRIPTION
**Pull request by stevenhao**
## Summary of changes

This PR refactors the boilerplate generator to generate the html string using javascript instead of using Blaze with Spacebars templates. Except for some minor differences (whitespace, html-escaping), the generated html is the same as it was before the refactor.

Changes to the `boilerplate-generator` package:
* Bump version to `1.1.2`
* Rename and modernize `boilerplate-generator.js` to `generator.js`
* Delete `boilerplate_web.browser.html` and `boilerplate_web.cordova.html`
* Add `template_web.browser.js` and `template_web.cordova.js`

This PR also adds a `boilerplate-generator-tests` package (previously, all the boilerplate-related tests were tested indirectly in the `webapp` package).

### CODE REVIEW

File: packages/boilerplate-generator/boilerplate-generator.js [hunk 1]
```diff
@@ -11,11 +11,9 @@ Boilerplate = function (arch, manifest, options) {
   options = options || {};
```
Comment (glasser) [inline packages/boilerplate-generator/boilerplate-generator.js]: This is really more of a question for @benjamn or @abernix but should new code be written with ES6 these days? import instead of Npm.require (it's ok even in packages to use import for built in packages like this one, right?), no var, class, etc? Using mainModule and imports instead of file lists and package-local variables?

And specifically should this refactored class be rewritten to new style or left alone?

File: packages/boilerplate-generator/boilerplate-generator.js [hunk 2]
```diff
@@ -27,12 +25,10 @@ Boilerplate = function (arch, manifest, options) {
...
   var self = this;
 
-  if (! self.baseData || ! self.func)
+  if (! self.baseData || ! self.template)
     throw new Error('Boilerplate did not instantiate correctly.');
 
-  return  "<!DOCTYPE html>\n" +
-    Blaze.toHTML(Blaze.With(_.extend(self.baseData, extraData),
-                            self.func));
+  return  "<!DOCTYPE html>\n" + self.template(_.extend(self.baseData, extraData));
```
Comment (glasser) [inline packages/boilerplate-generator/boilerplate-generator.js]: `_.extend` modifies self.baseData in place — is that intentional?  I think maybe `_.extend({}, self.baseData, extraData)` may be better. I see that's in the existing code too but maybe that's a bug.

File: packages/boilerplate-generator/boilerplate_web_template.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
+// Template function for rendering the boilerplate html
+// Replicates the template defined in boilerplate_web.browser.html
+// XXX Does not necessarily preserve formatting (e.g. additionalStaticJs newlines)
+// Arguments: root : { htmlAttributes, css : [{ url }], bundledJsCssUrlRewriteHook : Function, head, dynamicHead, body, dynamicBody, inlineScriptsAllowed, additionalStaticJs, meteorRuntimeConfig }
+
+Boilerplate_Web_Browser_Template = function(root) {
```
Comment (glasser) [inline packages/boilerplate-generator/boilerplate_web_template.js]: This isn't our normal variable naming style.  I think BoilerplateWebBrowserTemplate, or even just WebBrowserTemplate since we're not exporting it out of the package.

File: packages/boilerplate-generator/boilerplate_web_template.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  return [].concat(
+
+    // XXX is htmlAttributes ever anything but {}?
+    // may just be a generic Blaze/Spacebars thing.
+    function props1(htmlAttributes) {
```
Comment (glasser) [inline packages/boilerplate-generator/boilerplate_web_template.js]: I don't think you need to name these functions that are only called immediately. But I also am unsure why this is necessary — can't you just write root.htmlAttributes directly? Is this based on starting with some generated code?  I'd simplify this to look more like functions you'd write yourself.

File: packages/boilerplate-generator/boilerplate_web_template.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+      return [
+        ['<html'].concat(Object.keys(htmlAttributes).map(function(key) {
+
+          // XXX probably need to wrap strings in "".
```
Comment (glasser) [inline packages/boilerplate-generator/boilerplate_web_template.js]: Definitely want to have some escaping here! Not just wrapping, also doing proper encoding

File: packages/boilerplate-generator/boilerplate_web_template.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+      root.head,
+      root.dynamicHead,
+      '</head>',
+      '<body>',
+      root.body,
+      root.dynamicBody,
+      ''
+    ],
+
+    function if1(inlineScriptsAllowed) {
```
Comment (glasser) [inline packages/boilerplate-generator/boilerplate_web_template.js]: this one might be easier just as a ?: thing?

File: packages/boilerplate-generator/boilerplate_web_template.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+    [
+      '<head>'
+    ],
+
+    function each1(css) {
+      var bundledJsCssUrlRewriteHook = root.bundledJsCssUrlRewriteHook;
+      return _.map(css, function(obj) {
+        var url = obj.url;
+        return '  <link rel="stylesheet" type="text/css" class="__meteor-css__" href="' + bundledJsCssUrlRewriteHook(url) + '">';
```
Comment (glasser) [inline packages/boilerplate-generator/boilerplate_web_template.js]: url needs escaping (anything in the original template that's `{{}}` rather than `{{{}}}` should be escaped).  Note that escaping is either URL-escaping or HTML-escaping depending on where it is, I think?

File: packages/boilerplate-generator/package.js [hunk 8]
```diff
@@ -22,3 +25,16 @@ Package.onUse(function (api) {
     'boilerplate_web.cordova.html'
   ], 'server');
 });
+} else {
+
+Package.onUse(function (api) {
+  api.use([
+    'underscore',
```
Comment (glasser) [inline packages/boilerplate-generator/package.js]: Given that we're relying on underscore anyway, maybe just use [`_.template`](http://underscorejs.org/#template) with a template stored in a file and read by Assets.getText?

Comment (stevenhao) [inline packages/boilerplate-generator/package.js]: I had no idea this existed! I'll try it out.

File: packages/boilerplate-generator/boilerplate-generator.js (hunk 1 above)
Comment (benjamn) [inline packages/boilerplate-generator/boilerplate-generator.js]: Let's modernize this code! It runs on the server anyway.

File: packages/boilerplate-generator/boilerplate-generator.js (hunk 2 above)
Comment (benjamn) [inline packages/boilerplate-generator/boilerplate-generator.js]: Here's an idea you can ignore for the time being:

I would love to enable more interesting extensions of `self.baseData` here, specifically to support server-side rendering. For example, if this code called some custom callback with `self.baseData`, then the callback could parse the `self.baseData.body` HTML, attach server-rendered fragments to DOM nodes with particular `id`s. Then the client could use `document.getElementById` to find the same nodes and rehydrate the server-generated HTML.

For now, let's focus on feature parity, but I wanted to plant that seed anyway. 🌱

File: packages/boilerplate-generator/boilerplate_web_template.js (hunk 5 above)
Comment (stevenhao) [inline packages/boilerplate-generator/boilerplate_web_template.js]: Yeah, I've been delaying implementing this because I'm not sure what the best library for this kind of stuff. 

From what I can tell after reading https://github.com/meteor/blaze/blob/master/packages/htmljs/visitors.js#L198, https://html.spec.whatwg.org/multipage/syntax.html#attributes-2,

an attribute pair should stringify to `_.template(' <%= key %>="<%- value %>"')({key, value})`.

Potential issues: 
* The attribute namespace isn't explicitly handled and must be directly attached to the key of `htmlAttributes`, e.g. `htmlAttributes` may need to have entries like `{"xmlns:h": "http://www.w3.org/2000/xmlns/"}`.
* `_.template("<%- ... %>")` does HTML escaping, does not distinguish between *attribute mode*.  [this project](https://github.com/mathiasbynens/he) has an `isAttributeValue` option that may be useful. It seems to me that the attribute mode distinction isn't hugely important, and that most browsers handle unnecessarily escaped characters gracefully.

File: packages/boilerplate-generator/boilerplate-generator.js (hunk 2 above)
Comment (stevenhao) [inline packages/boilerplate-generator/boilerplate-generator.js]: yup, i'll change it to something like `_.extend(_.extend({}, self.baseData), extraData)`, or if it's safe to use ES6, `Object.assign({}, self.baseData, extraData)`

Comment (glasser) [inline packages/boilerplate-generator/boilerplate-generator.js]: ES6 is definitely fine in server code and I think in client code too.

File: packages/boilerplate-generator/boilerplate_web_template.js (hunk 5 above)
Comment (glasser) [inline packages/boilerplate-generator/boilerplate_web_template.js]: we chatted about this in person and decided that we need to HTML-encode attribute values but not names, and don't need to URL-encode anything.

File: packages/boilerplate-generator/boilerplate-generator.js [hunk 9]
```diff
@@ -1,55 +1,46 @@
-var fs = Npm.require('fs');
-var path = Npm.require('path');
+import fs from 'fs';
+import path from 'path';
```
Comment (benjamn) [inline packages/boilerplate-generator/boilerplate-generator.js]: I would be more specific:
```js
import { readFile } from "fs";
import { ??? } from "path";
```
It looks like you're not using `path` anywhere in this file, so I think you can remove that line.

File: packages/boilerplate-generator/boilerplate-generator.js [hunk 10]
```diff
@@ -1,55 +1,46 @@
...
+const readUtf8FileSync = filename => Meteor.wrapAsync(fs.readFile)(filename, 'utf8');
 
-  self._generateBoilerplateFromManifestAndSource(
-    manifest,
-    self.template,
-    options
-  );
-};
+export class Boilerplate {
+  constructor (arch, manifest, options = {}) {
```
Comment (benjamn) [inline packages/boilerplate-generator/boilerplate-generator.js]: We don't generally put a space after method names (before the `(`).

File: packages/boilerplate-generator/boilerplate-generator.js [hunk 11]
```diff
@@ -1,55 +1,46 @@
...
+  // by 'webapp' to specify data that is only known at request-time).
+  toHTML (extraData) {
+    if (!this.baseData || !this.template)
+      throw new Error('Boilerplate did not instantiate correctly.');
 
-  return  "<!DOCTYPE html>\n" +
-    Blaze.toHTML(Blaze.With(_.extend(self.baseData, extraData),
-                            self.func));
-};
+    return  "<!DOCTYPE html>\n" + this.template(Object.assign({}, this.baseData, extraData));
```
Comment (benjamn) [inline packages/boilerplate-generator/boilerplate-generator.js]: This is pretty long line that might be worth breaking after the `+`, as before.

File: packages/boilerplate-generator/boilerplate_web_cordova_template.js [hunk 12]
```diff
@@ -0,0 +1,70 @@
+// Template function for rendering the boilerplate html for cordova
+// Replicates the template defined in boilerplate_web.cordova.html
+// Arguments: root : { htmlAttributes, css : [{ url }], bundledJsCssUrlRewriteHook : Function, head, dynamicHead, body, dynamicBody, inlineScriptsAllowed, additionalStaticJs, meteorRuntimeConfig }
+
+export default function(manifest) {
```
Comment (benjamn) [inline packages/boilerplate-generator/boilerplate_web_cordova_template.js]: We do generally put a space after the `function` keyword (before the `(`) for anonymous functions.

File: packages/boilerplate-generator/boilerplate-generator.js [hunk 13]
```diff
@@ -88,19 +79,18 @@ Boilerplate.prototype._generateBoilerplateFromManifestAndSource =
...
     });
-    var boilerplateRenderCode = SpacebarsCompiler.compile(
-      boilerplateSource, { isBody: true });
-
-    // Note that we are actually depending on eval's local environment capture
-    // so that UI and HTML are visible to the eval'd code.
-    // XXX the template we are evaluating relies on the fact that UI is globally
-      // available.
-    global.UI = UI;
-    self.func = eval(boilerplateRenderCode);
```
Comment (benjamn) [inline packages/boilerplate-generator/boilerplate-generator.js]: Glad to see this `eval` go away!

File: packages/boilerplate-generator/boilerplate_web_browser_template.js [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+      })
+    )),
+
+    [
+      '', '',
+      '</body>',
+      '</html>'
+    ],
+
+    ['', '<!-- Generated for cordova by boilerplate-generator -->']
```
Comment (benjamn) [inline packages/boilerplate-generator/boilerplate_web_browser_template.js]: Should this be referring to Cordova, if this is `boilerplate_web_browser_template.js`?

File: packages/boilerplate-generator/boilerplate-generator.js [hunk 15]
```diff
@@ -88,19 +79,18 @@ Boilerplate.prototype._generateBoilerplateFromManifestAndSource =
...
+    this.baseData = boilerplateBaseData;
+  }
 };
 
-var _getTemplate = _.memoize(function (arch) {
-  var filename = 'boilerplate_' + arch + '.html';
-  return Assets.getText(filename);
+// Returns a template function that, when called, produces the boilerplate
+// html as a string.
+const _getTemplate = _.memoize(arch => {
```
Comment (glasser) [inline packages/boilerplate-generator/boilerplate-generator.js]: I don't think the memoize is necessary anymore now that the function isn't doing disk IO.

File: packages/boilerplate-generator/boilerplate_web_browser_template.js [hunk 16]
```diff
@@ -0,0 +1,70 @@
+// Template function for rendering the boilerplate html for browsers
+// Replicates the template defined in boilerplate_web.browser.html
+// Arguments: root : { htmlAttributes, css : [{ url }], bundledJsCssUrlRewriteHook : Function, head, dynamicHead, body, dynamicBody, inlineScriptsAllowed, additionalStaticJs, meteorRuntimeConfig }
+
+export default function (manifest) {
+  const root = manifest;
```
Comment (glasser) [inline packages/boilerplate-generator/boilerplate_web_browser_template.js]: Not sure why you're renaming the parameter instead of just giving it whatever name you want it to have.

File: packages/boilerplate-generator/boilerplate_web_browser_template.js [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+  // XXX do we need to do some validation on the properties of root?
+  return [].concat(
+    [
+      '<html' +_.map(root.htmlAttributes, (value, key) =>
+        _.template(' <%= attrName %>="<%- attrValue %>"')({
+          attrName: key,
```
Comment (glasser) [inline packages/boilerplate-generator/boilerplate_web_browser_template.js]: A little confusing to choose to call it "key" and then a line later call the same thing "name"

File: packages/boilerplate-generator/boilerplate_web_browser_template.js [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+      '</head>',
+      '<body>',
+      root.body,
+      root.dynamicBody,
+      '',
+      _.template(root.inlineScriptsAllowed
+        ? '  <script type="text/javascript"><%= contents %></script>'
+        : '  <script type="text/javascript" src="<%- src %>"></script>'
+      )({
+        contents: _.template('__meteor_runtime_config__ = JSON.parse(decodeURIComponent(<%= conf %>))')({
```
Comment (glasser) [inline packages/boilerplate-generator/boilerplate_web_browser_template.js]: Not sure why you're doing nested calls to _.template instead of just putting this in the first template. Also, it's kinda weird that you're providing both keys (contents/src) to either template... why not do

```
root.inlineScriptsAllowed ?
  _.template('template1', args1) : _.template('template2', args2)
```
?

Also in the old version, `meteorRuntimeConfig` was `{{}}` (HTML-escaped) though it probably doesn't matter due to the use of encodeURIComponent.

File: packages/boilerplate-generator/boilerplate_web_browser_template.js [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+          conf: root.meteorRuntimeConfig
+        }),
+        src: root.rootUrlPathPrefix + '/meteor_runtime_config.js'
```
Comment (glasser) [inline packages/boilerplate-generator/boilerplate_web_browser_template.js]: the second half here can just go in the template...

File: packages/boilerplate-generator/boilerplate_web_browser_template.js [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+    ],
+
+    _.map(root.js, ({url}) =>
+      _.template('  <script type="text/javascript" src="<%- src %>"></script>')({
+        src: root.bundledJsCssUrlRewriteHook(url)
+      })
+    ),
+
+    _.map(root.additionalStaticJs, ({contents, pathname}) => (
+      _.template(root.inlineScriptsAllowed
```
Comment (glasser) [inline packages/boilerplate-generator/boilerplate_web_browser_template.js]: similarly, this should probably be separate template calls

File: packages/boilerplate-generator/boilerplate_web_browser_template.js [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+        ? '  <script><%= contents %></script>'
```
Comment (glasser) [inline packages/boilerplate-generator/boilerplate_web_browser_template.js]: Looks like contents should actually be escaped.

File: packages/boilerplate-generator/boilerplate_web_cordova_template.js [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+      '  <script type="text/javascript" src="/cordova.js"></script>'
+    ],
+    _.map(root.js, ({url}) =>
+      _.template('  <script type="text/javascript" src="<%- src %>"></script>')({
+        src: url
+      })
+    ),
+    _.map(root.additionalStaticJs, ({pathname, contents}) =>
+      _.template(inlineScriptsAllowed
+        ? '  <script type="text/javascript"><%= contents %></script>'
```
Comment (glasser) [inline packages/boilerplate-generator/boilerplate_web_cordova_template.js]: escape, use multiple templates.

Although honestly I'm not sure why we check for inlineScriptsAllowed here when there's already a giant inline script above. Seems like a bug.  Not sure who actually knows about this stuff anymore — overlap of the browser content policy package and Cordova.  @benjamn ? @martijnwalraven ?

File: packages/boilerplate-generator/boilerplate_web_cordova_template.js [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+      root.head,
+      '</head>',
+      '',
+      '<body>',
+      root.body,
+      '</body>',
+      '</html>'
+    ],
+
+    ['', '<!-- Generated for cordova by boilerplate-generator -->']
```
Comment (glasser) [inline packages/boilerplate-generator/boilerplate_web_cordova_template.js]: Are these comments going to stick around forever?

File: packages/boilerplate-generator/boilerplate_web_browser_template.js (hunk 17 above)
Comment (stevenhao) [inline packages/boilerplate-generator/boilerplate_web_browser_template.js]: I agree that it's a bit confusing, but I was hoping that it would make sense:
- `(value, key)` correspond to `{key: value}` pairs
- `attrName` and `attrValue` are are templated as `name="value"` HTML attributes.

I'm happy to change it to be ``<% key %.="<%- value %>"'` if you think that's less confusing.

File: packages/boilerplate-generator/boilerplate_web_browser_template.js (hunk 21 above)
Comment (stevenhao) [inline packages/boilerplate-generator/boilerplate_web_browser_template.js]: It seems like we have to be careful with over-escaping `contents` here, because some browsers don't handle escaping very well when it's within a script tag.

[This Gist](https://gist.github.com/stevenhao/40f852353089e3d27391266eea6b67ba) shows how unexpected behavior can occur.

Comment (stevenhao) [inline packages/boilerplate-generator/boilerplate_web_browser_template.js]: Discussed this with @benjamn, we decided that it's probably safest to load all `additionalStaticJs` from the server. The reason is that embedding arbitrary javascript inline is tricky to do correctly, and can cause strange bugs.

The tradeoff we're making here is that loading from the server may be more costly than inlining the script. The only internal package that is loaded through `additionalStaticJs` is https://github.com/meteor/meteor/tree/master/packages/reload-safetybelt, so hopefully this change won't be too disruptive?

File: packages/boilerplate-generator/boilerplate_web_cordova_template.js (hunk 23 above)
Comment (stevenhao) [inline packages/boilerplate-generator/boilerplate_web_cordova_template.js]: I thought it might be nice to leave a note in generated code about how it was generated, but maybe it doesn't make sense to expose this to the public

File: packages/boilerplate-generator/boilerplate_web_cordova_template.js (hunk 22 above)
Comment (stevenhao) [inline packages/boilerplate-generator/boilerplate_web_cordova_template.js]: I'm going to err on the side of caution and not change the behavior of this section. If we decide that we'll be inlining all scripts for cordova, or inline no scripts for browser, (or both), I think that can be a separate PR.

File: packages/boilerplate-generator/boilerplate_web_browser_template.js (hunk 21 above)
Comment (glasser) [inline packages/boilerplate-generator/boilerplate_web_browser_template.js]: Hmm, I mean, this was an explicit choice to minimize round trips (ie, startup time) for apps, and I think it's been working for many years?  If it's only used by reload-safetybelt, a weird package we don't really even recommend, then I guess it's not a big deal, but in that case maybe it's not a big deal in either direction (if we are confident that reload-safetybelt's code is escaped correctly)?

**Comment by glasser** [comment]
Looking generally good. I'll leave what to do about additionalStaticJs up to Meteor team.  This is the last feedback I can give until back from vacation next Wednesday — I'm sure Meteor team can help you finish this one up :)

File: packages/boilerplate-generator/boilerplate_web_browser_template.js [hunk 24]
```diff
@@ -0,0 +1,76 @@
+// Template function for rendering the boilerplate html for browsers
```
Comment (abernix) [inline packages/boilerplate-generator/boilerplate_web_browser_template.js]: Certainly doesn't affect the outcome but I don't see a reason why we would abandon our previous naming convention on these new template files and not keeping the `arch` present in the filename verbatim. (For example, this file might be  `boilerplate_web.cordova_template.js` since the original was `boilerplate_web.cordova.html`).

File: packages/boilerplate-generator/boilerplate_web_cordova_template.js [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+      '  <script type="text/javascript">',
+      _.template('    __meteor_runtime_config__ = JSON.parse(decodeURIComponent(<%= conf %>));')({
+        conf: meteorRuntimeConfig
+      }),
+      '    if (/Android/i.test(navigator.userAgent)) {',
+      // When Android app is emulated, it cannot connect to localhost,
+      // instead it should connect to 10.0.2.2
+      // (unless we\'re using an http proxy; then it works!)
+      '      if (!__meteor_runtime_config__.httpProxyPort) {',
+      '        __meteor_runtime_config__.ROOT_URL = (__meteor_runtime_config__.ROOT_URL || \'\').replace(/localhost/i, \'10.0.2.2\');',
```
Comment (abernix) [inline packages/boilerplate-generator/boilerplate_web_cordova_template.js]: I know some linting algorithms or style-guides may unnecessarily chastise this, but on the Meteor project, feel free to switch to double quotes to avoid the excessive escaping of single-quotes. (i.e. to avoid having to `\'\'`)

File: packages/boilerplate-generator/boilerplate_web_cordova_template.js [hunk 26]
```diff
@@ -0,0 +1,79 @@
...
+    ),
+    [
+      '  <script type="text/javascript">',
+      _.template('    __meteor_runtime_config__ = JSON.parse(decodeURIComponent(<%= conf %>));')({
+        conf: meteorRuntimeConfig
+      }),
+      '    if (/Android/i.test(navigator.userAgent)) {',
+      // When Android app is emulated, it cannot connect to localhost,
+      // instead it should connect to 10.0.2.2
+      // (unless we\'re using an http proxy; then it works!)
```
Comment (abernix) [inline packages/boilerplate-generator/boilerplate_web_cordova_template.js]: We don't need to escape the `'` in this comment (it's not part of a string), but we should think about whether this comment was providing any value in the original template since it's no longer included now.  (I think the value is low, so mainly just pointing this out).

File: packages/boilerplate-generator-tests/cordova_test.js [hunk 27]
```diff
@@ -0,0 +1,90 @@
...
+});
+
+Tinytest.add("boilerplate-generator-tests - web.cordova include css", function (test) {
+  const html = generateHTML();
+  test.matches(html, /<link.*href=".*bootstrap.*">/);
+});
+
+Tinytest.add("boilerplate-generator-tests - web.cordova escape css", function (test) {
+  const html = generateHTML();
+  test.matches(html, /<link.*href=".*bootstrap.*&amp;v=&quot;1&quot;.*">/);
```
Comment (abernix) [inline packages/boilerplate-generator-tests/cordova_test.js]: Not that it would happen, but just pointing out that this will also match:

```html
<link rel="stylesheet" type="text/css" class="__meteor-css__" href="&v="1"bootstrap&amp;v=&quot;1&quot;">
```

Is that what you wanted?  Maybe better to test for the *absence* of quotes and `&` or for an exact attribute value match?  This might be easy to accomplish using `parse5`'s parse?

File: packages/boilerplate-generator-tests/cordova_test.js [hunk 28]
```diff
@@ -0,0 +1,90 @@
...
+});
+
+Tinytest.add("boilerplate-generator-tests - web.cordova do not call rewriteHook", function (test) {
```
Comment (abernix) [inline packages/boilerplate-generator-tests/cordova_test.js]: Just informational, but if you add another `-` after `web.cordova` in these test names, they will be grouped into their own section in tinytest reports. :)

File: packages/boilerplate-generator-tests/browser_test.js [hunk 29]
```diff
@@ -0,0 +1,100 @@
+import { parse, serialize } from 'parse5';
```
Comment (abernix) [inline packages/boilerplate-generator-tests/browser_test.js]: I think the name of this file should be consistent with that of the appropriate `arch` template (i.e. `boilerplate_web.browser.template.js`), and it should be plural like our other test files (e.g. `web.browser-tests.js`, `web.cordova-tests.js`)

File: packages/boilerplate-generator/boilerplate_web_browser_template.js [hunk 30]
```diff
@@ -0,0 +1,76 @@
...
+        : _.template('  <script type="text/javascript" src="<%- src %>/meteor_runtime_config.js"></script>')({
+          src: rootUrlPathPrefix
+        })
+      ) ,
+      ''
+    ],
+
+    _.map(js, ({url}) =>
+      _.template('  <script type="text/javascript" src="<%- src %>"></script>')({
+        src: bundledJsCssUrlRewriteHook(url) + '&hello=bye'
```
Comment (stevenhao) [inline packages/boilerplate-generator/boilerplate_web_browser_template.js]: ah, forgot to remove this from testing! thanks for catching it

**Comment by hwillson** [comment]
This is looking great @stevenhao! One small thing to mention (and it's not necessarily something that should be considered with these changes, but I'm just mentioning it for discussion). With these changes we're still forcing the Meteor generated CSS to come before any externally referenced CDN based CSS files (in `<head />`). For the full backstory, see https://github.com/meteor/meteor-feature-requests/issues/24 (and https://github.com/meteor/meteor/issues/8651). It's quite common practice to have 3rd party CDN based externally referenced CSS libraries come first in the CSS load order, which are then overridden as necessary by an application's CSS. Since we're refactoring the boilerplate it might be nice to also consider handling this scenario. Anyways, just food for thought - thanks for all of your work on this!

**Comment by stevenhao** [comment]
@hwillson, thanks for the comments! I agree that forcing the bundled css to come before custom css is not ideal, but I'm worried about backwards compatibility. Adding a configurable 
flag to reorder stylesheets is a possible solution, but that may be out of the scope of this PR. What do you think @abernix?

**Comment by abernix** [comment]
It's certainly worth mentioning the CSS load order here, but I think we should wrap this PR up and then iterate on it again as necessary.

This appears to be working properly in that when using `static-html` the server-side dependency on `blaze` is removed which was a goal of this PR (client was already removable).  It's worth pointing out that there is still a dependency on [`templating-tools`](https://github.com/meteor/blaze/tree/956709d7cad0f0c96c97796aa78e6124918a2e29/packages/templating-tools) (in the Blaze repo) because of:

1. The [`CompileError` class](https://github.com/meteor/blaze/blob/956709d7cad0f0c96c97796aa78e6124918a2e29/packages/templating-tools/templating-tools.js#L3) (via `throwCompileError`, [here](https://github.com/meteor/blaze/blob/956709d7cad0f0c96c97796aa78e6124918a2e29/packages/static-html/static-html.js#L87))
2. [`TemplatingTools.scanHtmlForTags`](https://github.com/meteor/blaze/blob/956709d7cad0f0c96c97796aa78e6124918a2e29/packages/static-html/static-html.js#L5).

I made a couple formatting changes but this is great work, @stevenhao!
